### DISCUSSION TITLE
Security Mode Command not Accepted by UE in 1.5.152

### DISCUSSION THREAD

**Discussion by shehzadalee** [magma/magma#8060] [Q&A]
An IMSI/UE connected to AGW Version 1.3.3 gets normally connected while The same UE/IMSI when sends an attach to AGW bc13 1.5.152 gets Security Mode Reject.
Comparing the UE Capabilities with that of security algorithms and they match 100% between both the cases except that in 1.5.152 we also send Encryption/Integrity algorithms for GPRS in addition to that of LTE and UMTS. (which I dont think should cause an issue).
However, the KASME Value that is computed at MME based on Kc and Ki values received from Sigscale HSS. Based on those values MME computes the KASME which is sent in Security Mode Command message to the UE. Doubting if that may be an issue? How can we resolve it?
[Security mode Command.zip](https://github.com/magma/magma/files/6824549/Security.mode.Command.zip)

**Comment by amarpad** [comment]
cc @ulaskozat @pruthvihebbani @rsarwad

**Reply by wallyrb** [reply]
@shehzadalee what mme logs are showing in the `Security Mode Reject` , could you share these logs?

**Reply by shehzadalee** [reply]
Hi @wallyrb Please find attached as requested.
[Logs.zip](https://github.com/magma/magma/files/6910499/Logs.zip)

**Comment by jpad-freedomfi** [comment]
@wallyrb We have update on this topic. 

We still see the behavior in attached logs. Here we have AGW on 1.5 branch head, feg at 1.5.84, and ocr8r at 1.5. Still see MAC failure. 
With same setup, if we revert AGW to 1.3.2, same UE works. 

Logs:
mme
syslog
s1 pcap
s6a pcap
[Mac_failure_dishnet.zip](https://github.com/magma/magma/files/7003452/Mac_failure_dishnet.zip)

**Reply by ulaskozat** [reply]
@amarpad @jpad-freedomfi I am just seeing this thread of discussion. Do you see this issue for other UEs too?

**Reply by ulaskozat** [reply]
on s6a pcap, I do not see any diameter messages on wireshark though. Are we sure that this UE's s6a traffic is going to FEG?

**Reply by ulaskozat** [reply]
Though in the log I can see the correct mod selection. In terms of message formatting, etc. in Authentication Request, I do not see anything standing out. Can we verify if the Authentication Vector on S1 is actually matching what HSS is sending on S6a?

**Reply by jpad-freedomfi** [reply]
@ulaskozat Diameter is on port 3868 and 3869. You'll need to update preferences in Wireshark to see the 3868 packets.

**Reply by ulaskozat** [reply]
@jpad-freedomfi thanks 3869 was not listed on diameter. The AUTN value does not match on these S6a and S1 pcaps. However the timestamps on S6a and S1 are different.

**Comment by jpad-freedomfi** [comment]
@ulaskozat we made progress on this last night. Baicells eNB was configured for EIA1. 

With 1.3, we can leave the default MME config (EIA2, EIA1, EIA0) and it works. With 1.5 we have to setup EIA1 as first preference. Then security mode completes. 

I haven't had this issue before. Unclear to me what changed between 1.3 and 1.5 to make the selection process different and/or what role the eNB configuration is playing in this.

**Reply by ulaskozat** [reply]
@jpad-freedomfi this is surprising as the failing step is the Authentication and the integrity algo etc. selection happens during SMC.

**Reply by jpad-freedomfi** [reply]
MAC calculation uses EIA. Not sure how the EIA version for this is selected. This is first time I have noticed it set in eNB to specific EIA version. Baicells radio.

**Reply by ulaskozat** [reply]
@jpad-freedomfi this MAC is part of the authentication vector, it is not the MAC generated for NAS payload. It comes from S6a.

**Reply by ulaskozat** [reply]
IA and EA selection happens during SMC step (which does not happen here due to authentication failure). The selected algorithm is communicated in SMC.

**Comment by shehzadalee** [comment]
Hi @ulaskozat 

We are still observing this issue with Sigscale HSS. Attached are some fresh captures for the same IMSI which gets successfully attached using Magma HSS but Security mode Procedure fails while using Sigscale HSS.

We tried to modify Integrity order at Magma and eNodeB but no combination worked which was successful previously when the order on Magma was changed.

[SMC.zip](https://github.com/magma/magma/files/7048083/SMC.zip)

Note that this happens with all the UEs trying to attach in this environment.

Also, Is there any way to capture S6a captures on GRPC?

**Reply by ulaskozat** [reply]
you should be able to see them on FEG. @uri200 can you point to @shehzadalee on how to enable grpc monitoring on the FEG?

**Reply by jpad-freedomfi** [reply]
oh, good point. We know how to enable grpc monitoring. We'll do that. Thanks @ulaskozat

**Reply by ulaskozat** [reply]
@jpad-freedomfi  since you have the pcaps, I am not sure what other useful info we can get from grpc messages additionally. On pcaps, I can see the authentication vectors are matching with Sigscale HSS (if they were not, authentication step would fail anyhow). Since UE is not sending a specific reason for SMC reject, it is hard to say what it is complaining about.

**Reply by ulaskozat** [reply]
As I mentioned earlier, I do not see the relevance of reordering EIA preferences in the problems you are observing. The behavior looks random too me on part of the UE. It may be a pure luck that without Sigscale HSS it is going through. 

Are other UEs happy?

**Reply by ulaskozat** [reply]
Sorry I see from @shehzadalee's response that all the UEs observe the same outcome. Are they all failing at the same step?

**Reply by shehzadalee** [reply]
@ulaskozat Yes they all fail at the same step.

**Reply by ulaskozat** [reply]
On the SMC from PCAPs, everything looks the same with the expected differences in MAC and s1ap identifiers.

**Reply by jpad-freedomfi** [reply]
Does MAC have a nonce or any temporal or salted data? Or should it match?

**Reply by ulaskozat** [reply]
MAC for the payload is specific to the payload and derived integration key.

**Comment by ulaskozat** [comment]
cc @lionelgo, do you have more insights on this one?

**Reply by lionelgo** [reply]
I would have to check with the 1.5 release but in mme logs, the following seems suspucious to me:
000346 Sat Jul 31 00:11:45 2021 7F3844085700 INFO  NAS-EM tasks/nas/emm/sap/emm_cn.c      :1138               EMMCN-SAP - Received primitive EMM_CN_AUTHENTICATION_PARAM_RES (401)
000347 Sat Jul 31 00:11:45 2021 7F3844085700 INFO  NAS-EM tasks/nas/emm/sap/emm_fsm.c     :0241                              EMM-FSM   - Received event COMMON_PROC_ABORT (4) in state EMM-DEREGISTERED
000348 Sat Jul 31 00:11:45 2021 7F3844085700 **ERROR** NAS-EM tasks/nas/emm/sap/EmmDeregistere:0123                                 EMM-FSM state EMM_DEREGISTERED - Primitive ** _EMMREG_COMMON_PROC_ABORT** is not valid

**Reply by lionelgo** [reply]
I checked the replayed UE sec cap, did not find something.
Then if nothing more pop up, I would check if the integrity of SMC match the EIA(2) selected.

**Reply by lionelgo** [reply]
Do you run AGWs in VMs or do you run both on same physical host ?

**Reply by shehzadalee** [reply]
Hey @lionelgo, on same Physical AGW in both the scenarios

**Reply by lionelgo** [reply]
Just in case this reminds somebody about something, I made a diff between the 2 releases,I found S1AP rel 10 versus 15, only EIA1 code have difference, in SMC gea bit handling algorithm have changed but result is fine, I checked that nas_stream_encrypt_eia2() use the same lib/version.
What I can do tomorrow is testing release 1.5.2 with dsTester, or better 1.5.152 if relevant, but  I don't know how to get this last one, if somebody can just point me how to get it.

**Reply by ulaskozat** [reply]
@shehzadalee 1.5.152 is FFI build, right? can you share the mme binary with Lionel? @lionelgo note that they are facing the issue in federated mode only.

**Reply by shehzadalee** [reply]
Hey @ulaskozat 
Thats true its FFi build, let me share the binary with you.

**Reply by shehzadalee** [reply]
Please find the binary at below google drive link
https://drive.google.com/file/d/1J5IezJawU56ZGdJEz59YneZM_337eTVp/view?usp=sharing

**Reply by shehzadalee** [reply]
Hi @lionelgo 
Were you able to access the binary?

**Comment by jpad-freedomfi** [comment]
@ulaskozat as a data point, we had this happen with a second customer installation. 1.5 + Sigscale HSS, UEs would not get passed Security mode, when we rolled back to 1.3, UEs attach right away. I think next week will try to setup side by side networks. If you can think of logs/captures that would help we can get them.

**Comment by blueCARDSS** [comment]
hi 

we setup lab and discover UE was rejected due to NAS Mobility  Magmagment Message Type : Attach rejected 
Cause : Network Failure  
AGW Version : 1.5.193-1628512814-2f819bb773 
![image_2021_10_28T11_09_11_677Z (1)](https://user-images.githubusercontent.com/92744847/139258939-2959e702-ac63-4a67-adf3-e16fa8eda9a1.png)
![image_2021_10_28T11_09_29_582Z](https://user-images.githubusercontent.com/92744847/139258962-0fee0408-962d-4bb2-99a8-14669a559b4d.png)

**Comment by crbertoldo** [comment]
Hi @ulaskozat do you know whether this issue was somehow addressed?

**Reply by ulaskozat** [reply]
Please reach out to Radisys team (cc @rsarwad).

**Reply by rsarwad** [reply]
Fine @ulaskozat I will look into the issue. 
@crbertoldo, on which branch you notice this issue recently?

**Reply by crbertoldo** [reply]
Hi @rsarwad, thank you. We are working with the branch v1.6.

**Reply by crbertoldo** [reply]
Hey @rsarwad, the HSS vendor said R values were encrypted is DES and not AES, then after changing it, we could move on. Thank you!

**Reply by rsarwad** [reply]
Thank you, @crbertoldo for letting us know.
